### ISSUE TITLE
Handling NULL values in SQL

### ISSUE THREAD

**Issue by tracykteal** [datacarpentry/sql-ecology-lesson#35]
The dataset we use has a lot of empty cells, so the issue of how to handle missing data often comes up. Can we add a section on handling missing data, like with IS NULL and IS NOT NULL?

**Comment by tpoisot** [comment]
It looks like the firefox extension puts empty fields, instead of `NULL`, when importing. I'll look into it.

**Comment by waveform80** [comment]
Hmm, just had a go importing [`surveys.csv`](https://figshare.com/articles/Portal_Project_Teaching_Database/1314459) into the `surveys` table after emptying it (with `DELETE FROM surveys`), using the firefox extension. Seems to insert NULLs for me:

```sql
SELECT COUNT(*)
FROM surveys WHERE species_id IS NULL;
```

The query above yields 763 as the result after the import (which is what was in the original database). For reference I'm using SQLite manager 0.8.3.1 on Firefox 45.0.

**Comment by tpoisot** [comment]
Referenced in https://github.com/swcarpentry/sql-novice-survey/issues/135 too.

**Comment by waveform80** [comment]
swcarpentry/sql-novice-survey#135 seems to be dealing with the results of the `.import` command which does import empty fields as blank strings (as the explanation in the [linked thread](http://sqlite.1065341.n5.nabble.com/import-null-values-td3405.html) details, `.import` binds everything as a text value and leaves it to the SQL layer to deal with data-type conversions). So the only method of getting NULLs from `.import` is as suggested in that ticket: fix-up the data with `UPDATE` afterward.

The firefox extension seems to be different in that it assumes all empty fields in a CSV file are NULL. I altered the `species_id` column to include a `NOT NULL` constraint and retried the import to see if it'd be smart enough to use a blank string instead when a `NOT NULL` constraint was present but, no, it's not (the import simply fails).

I then used the firefox extension to export the surveys table after modifying `species_id` to replace all NULLs with a blank string instead. The resulting file contains "" for blank strings instead of a purely empty field (a fairly common, but by no means universal, practice in CSV handling amongst database interfaces - unfortunately CSV has no universally accepted representation for NULL vs the blank string).

Re-importing that file after clearing `surveys` showed the firefox extension will treat "" as a blank string rather than NULL for import purposes too.

Given the lesson currently uses the firefox extension for importing, I don't think this is a serious problem, but it's probably worth pointing out that other import mechanisms may treat empty fields and blank strings differently and that one should check the results of an import before relying upon it.
